include/rsp_queue.inc

```diff
@@ -0,0 +1,566 @@
+.endm
+
+# Globally reserved register. Do not use it in overlay code
+# except to read command arguments.
+#define rspq_dmem_buf_ptr gp
+
+# Can be used with l* instructions to get contents of the current command at the specified offset.
+# The total command size needs to be specified as well.
+#define CMD_ADDR(offset, cmdsize) (%lo(RSPQ_DMEM_BUFFER) + (offset) - (cmdsize)) (rspq_dmem_buf_ptr)
+
+########################################################
+#
+# The following is the actual implementation of the rsp engine.
+# It is included in every overlay to verify binary
+# compatibility at runtime during overlay registration. 
+# But only the overlay code will actually be loaded into 
+# the RSP during execution.
+#
+########################################################
+
+#define OVERLAY_HEADER_SIZE 0x8
+
+    .set noreorder
+    .set at
+
+    .data
+_data_start:
+
+# Overlay tables. See rsp_overlay_t in rsp.c
+RSPQ_OVERLAY_TABLE:           .ds.b RSPQ_OVERLAY_TABLE_SIZE
+RSPQ_OVERLAY_DESCRIPTORS:     .ds.b (RSPQ_OVERLAY_DESC_SIZE * RSPQ_MAX_OVERLAY_COUNT)
+
+# Save slots for RDRAM addresses used during nested lists calls.
+# Notice that the two extra slots are used to save the lowpri
+# and highpri current pointer (used when switching between the two)
+RSPQ_POINTER_STACK:           .ds.l (RSPQ_MAX_BLOCK_NESTING_LEVEL+2)
+
+# RDRAM address of the current command list.
+RSPQ_RDRAM_PTR:               .long 0
+
+# Index (not ID!) of the current overlay, as byte offset in the descriptor array
+RSPQ_CURRENT_OVL:             .half 0
+
+    .align 4
+    .ascii "Dragon RSP Queue"
+    .ascii "Rasky & Snacchus"
+
+    .align 3
+RSPQ_INTERNAL_COMMAND_TABLE:
+RSPQ_DefineCommand RSPQCmd_WaitNewInput,    0     # 0x00
+RSPQ_DefineCommand RSPQCmd_Noop,            4     # 0x01
+RSPQ_DefineCommand RSPQCmd_Jump,            4     # 0x02
+RSPQ_DefineCommand RSPQCmd_Call,            8     # 0x03
+RSPQ_DefineCommand RSPQCmd_Ret,             4     # 0x04
+RSPQ_DefineCommand RSQPCmd_Dma,             16    # 0x05
+RSPQ_DefineCommand RSPQCmd_WriteStatus,     4     # 0x06 -- must be even (bit 24 must be 0)
+RSPQ_DefineCommand RSPQCmd_SwapBuffers,     12    # 0x07
+RSPQ_DefineCommand RSPQCmd_TestWriteStatus, 8     # 0x08 -- must be even (bit 24 must be 0)
+
+#if RSPQ_DEBUG
+RSPQ_LOG_IDX:                .long 0
+RSPQ_LOG:                    .ds.l 16
+RSPQ_LOG_END:                .long 0xFFFFFFFF
+#endif
+
+    .align 3
+RSPQ_DMEM_BUFFER:            .ds.b RSPQ_DMEM_BUFFER_SIZE
+
+
+    .align 4
+# Overlay data will be loaded at this address
+_ovl_data_start:
+
+    .text
+
+    .globl _start
+_start:
+    li rspq_dmem_buf_ptr, 0
+
+    .func RSPQCmd_WaitNewInput
+RSPQCmd_WaitNewInput:
+    # Check if new commands were added in the display list (SIG_MORE)
+    mfc0 t0, COP0_SP_STATUS
+    andi t0, SP_STATUS_SIG_MORE
+    bnez t0, wakeup
+    li t0, SP_WSTATUS_CLEAR_SIG_MORE
+
+    # No new commands yet, go to sleep
+    break
+wakeup:
+    mtc0 t0, COP0_SP_STATUS
```

Comment:
This resets the `SP_STATUS_SIG_MORE` signal (`SP_WSTATUS_CLEAR_SIG_MORE` is loaded into `t0` a couple of lines above), so we'll know if the CPU sets that signal in the time between the RSP waking up and checking the signal the next time. If it's still unset when checking again, the RSP goes back to sleep. Otherwise, it keeps fetching new commands. If we didn't reset the signal here, there would be cases where the RSP unnecessarily fetches new commands even though none are available. That wouldn't break anything, but would waste some DMA bandwidth.

Reply:
Ok this is where we clear the signal, I totally missed that. So is this example flow correct?

- `RSPQCmd_WaitNewInput` does the `mfc0 t0, COP0_SP_STATUS` and sees `SIG_MORE` is not set
- CPU writes `SIG_MORE` and starts waiting for 10 instructions
- RSP goes into sleep
- CPU re-writes `SIG_MORE`
- RSP wakes and clears `SIG_MORE`
- RSP executes the written command and the command jumps to `RSPQ_Loop` until there is no `RSPQCmd_WaitNewInput`
- RSP does the `mfc0 t0, COP0_SP_STATUS` sees his cleared value (i.e CPU didn't write anything and did not flush again) and sleeps

Reply:
Yes, exactly!